consensus/tendermint/backend/engine_test.go

```diff
@@ -453,22 +458,253 @@ func TestClose(t *testing.T) {
 		b := &Backend{}
 
 		err := b.Close()
-		if err != ErrStoppedEngine {
-			t.Fatalf("expected %v, got %v", ErrStoppedEngine, err)
-		}
+		assertNotCoreStarted(t, b)
+		assertError(t, ErrStoppedEngine, err)
 	})
 
 	t.Run("engine is running, no errors", func(t *testing.T) {
+		ctrl := gomock.NewController(t)
+		defer ctrl.Finish()
+
+		tendermintC := tendermintCore.NewMockTendermint(ctrl)
+		tendermintC.EXPECT().Stop().MaxTimes(1)
+
 		b := &Backend{
+			core:        tendermintC,
 			coreStarted: true,
 			stopped:     make(chan struct{}),
 		}
 
 		err := b.Close()
-		if err != nil {
-			t.Fatalf("expected <nil>, got %v", err)
+		assertNotCoreStarted(t, b)
+		assertNilError(t, err)
+	})
+
+	t.Run("engine is running, stopped twice", func(t *testing.T) {
+		ctrl := gomock.NewController(t)
+		defer ctrl.Finish()
+
+		tendermintC := tendermintCore.NewMockTendermint(ctrl)
+		tendermintC.EXPECT().Stop().MaxTimes(1)
+
+		b := &Backend{
+			core:        tendermintC,
+			coreStarted: true,
+			stopped:     make(chan struct{}),
 		}
+
+		err := b.Close()
+		assertNotCoreStarted(t, b)
+		assertNilError(t, err)
```

Comment:
These two statements look the wrong way around. `Close` returning an error indicates that something went wrong, at that point we shouldn't make any assumptions about what the effect the method call might have had. Here we are asserting that core was stopped irrespective of whether the method call succeeded.

Reply:
This is the expected behaviour since the engine is running when we call close, and it manages to close correctly then I am not expecting an error thus `assertNilError` check.

 The lines below call close on the already closed engine, so I expect an error to be returned thus `assertError` check.

Reply:
Sorry I wasn't clear. I can see that we are checking expected behaviour. Im just suggesting re-ordering these 2 lines. like so. And performing the same re-ordering on the 2 lines below.
```suggestion
		assertNilError(t, err)
		assertNotCoreStarted(t, b)
```

In this way if `Close` unexpectedly returns an error we will fail the test for that reason, whereas currently the test would fail because core would not be started but it would not report the error which would be the root cause.